File: backend/app/parsers/a1111.py
import re
from typing import Any

from app.parsers.base import (
    LoraInfo,
    MetadataParser,
    ParsedMetadata,
    SourceTool,
)
from app.parsers.model_detector import detect_model_type


class A1111Parser(MetadataParser):
    """Parser for A1111/Forge metadata format."""

    # Regex for LoRA tags: <lora:name:weight> or <lora:name:weight:clip_weight>
    LORA_PATTERN = re.compile(r"<lora:([^:>]+):([0-9.]+)(?::([0-9.]+))?>")

    def can_parse(self, png_info: dict[str, Any]) -> bool:
        """Check if the PNG info contains A1111/Forge metadata."""
        if "parameters" not in png_info:
            return False

        params = png_info["parameters"]
        if not isinstance(params, str):
            return False

        # Check for characteristic patterns
        return "Steps:" in params and "Sampler:" in params

    def parse(self, png_info: dict[str, Any]) -> ParsedMetadata:
        """Parse A1111/Forge metadata."""
        params_str = png_info.get("parameters", "")

        # Detect if it's Forge
        source_tool = SourceTool.A1111
        if "Forge" in params_str:
            source_tool = SourceTool.FORGE

        metadata = ParsedMetadata(
            source_tool=source_tool,
            has_metadata=True,
            raw_metadata={"parameters": params_str},
        )

        # Parse the parameters string
        self._parse_parameters(params_str, metadata)

        # Detect model type
        metadata.model_type = detect_model_type(metadata.model_name)

        return metadata

    def _parse_parameters(self, params_str: str, metadata: ParsedMetadata) -> None:
        """Parse the parameters string."""
        lines = params_str.strip().split("\n")

        # Find the line with "Steps:" - this separates prompts from params
        params_line_idx = -1
        for i, line in enumerate(lines):
            if line.startswith("Steps:"):
                params_line_idx = i
                break

        if params_line_idx == -1:
            return

        # Everything before the params line is prompts
        prompt_lines = lines[:params_line_idx]
        params_line = lines[params_line_idx]

        # Parse prompts
        self._parse_prompts(prompt_lines, metadata)

        # Parse parameters line
        self._parse_params_line(params_line, metadata)

        # Extract LoRAs from positive prompt
        if metadata.positive_prompt:
            self._extract_loras(metadata.positive_prompt, metadata)

    def _parse_prompts(self, lines: list[str], metadata: ParsedMetadata) -> None:
        """Parse positive and negative prompts from lines."""
        if not lines:
            return

        prompt_parts: list[str] = []
        negative_parts: list[str] = []
        in_negative = False

        for line in lines:
            if line.startswith("Negative prompt:"):
                in_negative = True
                # Get the part after "Negative prompt:"
                neg_text = line[len("Negative prompt:"):].strip()
                if neg_text:
                    negative_parts.append(neg_text)
            elif in_negative:
                negative_parts.append(line)
            else:
                prompt_parts.append(line)

        positive = " ".join(prompt_parts).strip()
        negative = " ".join(negative_parts).strip()

        if positive:
            metadata.positive_prompt = positive
        if negative:
            metadata.negative_prompt = negative

    def _parse_params_line(self, params_line: str, metadata: ParsedMetadata) -> None:
        """Parse the parameters line (key: value, key: value, ...)."""
        # Split by comma, but be careful with values that contain commas
        parts = self._split_params(params_line)

        model_params: dict[str, Any] = {}

        for part in parts:
            if ":" not in part:
                continue

            key, value = part.split(":", 1)
            key = key.strip()
            value = value.strip()

            # Map to metadata fields
            if key == "Steps":
                try:
                    metadata.steps = int(value)
                except ValueError:
                    pass
            elif key == "Sampler":
                metadata.sampler_name = value
            elif key == "CFG scale":
                try:
                    metadata.cfg_scale = float(value)
                except ValueError:
                    pass
            elif key == "Seed":
                try:
                    metadata.seed = int(value)
                except ValueError:
                    pass
            elif key == "Model":
                metadata.model_name = value
            elif key == "Scheduler":
                metadata.scheduler = value
            elif key == "Clip skip":
                try:
                    model_params["clip_skip"] = int(value)
                except ValueError:
                    pass
            elif key == "VAE":
                model_params["vae"] = value
            elif key == "Model hash":
                model_params["model_hash"] = value
            elif key == "Size":
                model_params["size"] = value
            # Hires upscaler parameters
            elif key == "Hires upscale":
                try:
                    model_params["hires_upscale"] = float(value)
                except ValueError:
                    pass
            elif key == "Hires upscaler":
                model_params["hires_upscaler"] = value
            elif key == "Hires steps":
                try:
                    model_params["hires_steps"] = int(value)
                except ValueError:
                    pass
            elif key == "Denoising strength":
                try:
                    model_params["denoising_strength"] = float(value)
                except ValueError:
                    pass
            # XYZ grid parameters
            elif key == "X/Y/Z plot":
                model_params["is_xyz_grid"] = True
            elif key == "X Type":
                model_params["xyz_x_type"] = value
            elif key == "X Values":
                model_params["xyz_x_values"] = value.strip('"')
            elif key == "Y Type":
                model_params["xyz_y_type"] = value
            elif key == "Y Values":
                model_params["xyz_y_values"] = value.strip('"')
            elif key == "Z Type":
                model_params["xyz_z_type"] = value
            elif key == "Z Values":
                model_params["xyz_z_values"] = value.strip('"')

        if model_params:
            metadata.model_params = model_params

    def _split_params(self, params_line: str) -> list[str]:
        """Split parameters line handling quoted values."""
        parts: list[str] = []
        current = ""
        in_quotes = False

        for char in params_line:
            if char == '"':
                in_quotes = not in_quotes
                current += char
            elif char == "," and not in_quotes:
                if current.strip():
                    parts.append(current.strip())
                current = ""
            else:
                current += char

        if current.strip():
            parts.append(current.strip())

        return parts

    def _extract_loras(self, prompt: str, metadata: ParsedMetadata) -> None:
        """Extract LoRA information from prompt text."""
        matches = self.LORA_PATTERN.findall(prompt)

        for match in matches:
            name = match[0]
            weight = float(match[1]) if match[1] else 1.0
            weight_clip = float(match[2]) if match[2] else None

            lora_info = LoraInfo(
                name=name,
                weight=weight,
                weight_clip=weight_clip,
            )
            metadata.loras.append(lora_info)

File: backend-rs/src/parser/a1111.rs
//! A1111 / Forge parameters-string parser (mirror parsers/a1111.py).

use std::collections::HashMap;
use std::sync::OnceLock;

use regex::Regex;
use serde_json::json;

use super::{detect_model_type, LoraInfo, ParsedMetadata, SourceTool};

/// `<lora:name:weight>` or `<lora:name:weight:clip_weight>`.
fn lora_re() -> &'static Regex {
    static RE: OnceLock<Regex> = OnceLock::new();
    RE.get_or_init(|| Regex::new(r"<lora:([^:>]+):([0-9.]+)(?::([0-9.]+))?>").unwrap())
}

pub fn can_parse(png_info: &HashMap<String, String>) -> bool {
    match png_info.get("parameters") {
        Some(p) => p.contains("Steps:") && p.contains("Sampler:"),
        None => false,
    }
}

pub fn parse(png_info: &HashMap<String, String>) -> ParsedMetadata {
    let params_str = png_info.get("parameters").cloned().unwrap_or_default();
    let source_tool = if params_str.contains("Forge") {
        SourceTool::Forge
    } else {
        SourceTool::A1111
    };
    let mut m = ParsedMetadata::new(source_tool);
    m.raw_metadata = Some(json!({ "parameters": params_str }));

    parse_parameters(&params_str, &mut m);
    m.model_type = Some(detect_model_type(m.model_name.as_deref()));
    m
}

fn parse_parameters(params_str: &str, m: &mut ParsedMetadata) {
    let lines: Vec<&str> = params_str.trim().split('\n').collect();

    // The "Steps:" line separates the prompt block from the parameters block.
    let Some(idx) = lines.iter().position(|l| l.starts_with("Steps:")) else {
        return;
    };

    parse_prompts(&lines[..idx], m);
    parse_params_line(lines[idx], m);

    if let Some(prompt) = m.positive_prompt.clone() {
        extract_loras(&prompt, m);
    }
}

fn parse_prompts(lines: &[&str], m: &mut ParsedMetadata) {
    let mut positive_parts: Vec<String> = Vec::new();
    let mut negative_parts: Vec<String> = Vec::new();
    let mut in_negative = false;

    for &line in lines {
        if let Some(rest) = line.strip_prefix("Negative prompt:") {
            in_negative = true;
            let neg = rest.trim();
            if !neg.is_empty() {
                negative_parts.push(neg.to_string());
            }
        } else if in_negative {
            negative_parts.push(line.to_string());
        } else {
            positive_parts.push(line.to_string());
        }
    }

    let positive = positive_parts.join(" ").trim().to_string();
    let negative = negative_parts.join(" ").trim().to_string();
    if !positive.is_empty() {
        m.positive_prompt = Some(positive);
    }
    if !negative.is_empty() {
        m.negative_prompt = Some(negative);
    }
}

fn parse_params_line(params_line: &str, m: &mut ParsedMetadata) {
    for part in split_params(params_line) {
        let Some((key, value)) = part.split_once(':') else {
            continue;
        };
        let key = key.trim();
        let value = value.trim();
        let mp = &mut m.model_params;
        match key {
            "Steps" => {
                if let Ok(v) = value.parse::<i32>() {
                    m.steps = Some(v);
                }
            }
            "Sampler" => m.sampler_name = Some(value.to_string()),
            "CFG scale" => {
                if let Ok(v) = value.parse::<f64>() {
                    m.cfg_scale = Some(v);
                }
            }
            "Seed" => {
                if let Ok(v) = value.parse::<i64>() {
                    m.seed = Some(v);
                }
            }
            "Model" => m.model_name = Some(value.to_string()),
            "Scheduler" => m.scheduler = Some(value.to_string()),
            "Clip skip" => {
                if let Ok(v) = value.parse::<i64>() {
                    mp.insert("clip_skip".into(), json!(v));
                }
            }
            "VAE" => {
                mp.insert("vae".into(), json!(value));
            }
            "Model hash" => {
                mp.insert("model_hash".into(), json!(value));
            }
            "Size" => {
                mp.insert("size".into(), json!(value));
            }
            "Hires upscale" => {
                if let Ok(v) = value.parse::<f64>() {
                    mp.insert("hires_upscale".into(), json!(v));
                }
            }
            "Hires upscaler" => {
                mp.insert("hires_upscaler".into(), json!(value));
            }
            "Hires steps" => {
                if let Ok(v) = value.parse::<i64>() {
                    mp.insert("hires_steps".into(), json!(v));
                }
            }
            "Denoising strength" => {
                if let Ok(v) = value.parse::<f64>() {
                    mp.insert("denoising_strength".into(), json!(v));
                }
            }
            "Script" if value == "X/Y/Z plot" => {
                mp.insert("is_xyz_grid".into(), json!(true));
            }
            "X Type" => {
                mp.insert("xyz_x_type".into(), json!(value));
            }
            "X Values" => {
                mp.insert("xyz_x_values".into(), json!(value.trim_matches('"')));
            }
            "Y Type" => {
                mp.insert("xyz_y_type".into(), json!(value));
            }
            "Y Values" => {
                mp.insert("xyz_y_values".into(), json!(value.trim_matches('"')));
            }
            "Z Type" => {
                mp.insert("xyz_z_type".into(), json!(value));
            }
            "Z Values" => {
                mp.insert("xyz_z_values".into(), json!(value.trim_matches('"')));
            }
            _ => {}
        }
    }
}

/// Split a parameters line on commas, ignoring commas inside double quotes.
fn split_params(line: &str) -> Vec<String> {
    let mut parts = Vec::new();
    let mut current = String::new();
    let mut in_quotes = false;
    for ch in line.chars() {
        match ch {
            '"' => {
                in_quotes = !in_quotes;
                current.push(ch);
            }
            ',' if !in_quotes => {
                let t = current.trim();
                if !t.is_empty() {
                    parts.push(t.to_string());
                }
                current.clear();
            }
            _ => current.push(ch),
        }
    }
    let t = current.trim();
    if !t.is_empty() {
        parts.push(t.to_string());
    }
    parts
}

fn extract_loras(prompt: &str, m: &mut ParsedMetadata) {
    for caps in lora_re().captures_iter(prompt) {
        let name = caps[1].to_string();
        let weight = caps.get(2).and_then(|w| w.as_str().parse().ok()).unwrap_or(1.0);
        let weight_clip = caps.get(3).and_then(|w| w.as_str().parse().ok());
        m.loras.push(LoraInfo {
            name,
            weight,
            weight_clip,
            hash: None,
        });
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parses_basic_a1111() {
        let params = "masterpiece, 1girl <lora:detail:0.8>\n\
            Negative prompt: bad hands\n\
            Steps: 28, Sampler: DPM++ 2M, CFG scale: 7.0, Seed: 12345, Model: animagineXL";
        let mut info = HashMap::new();
        info.insert("parameters".to_string(), params.to_string());
        assert!(can_parse(&info));
        let m = parse(&info);
        assert_eq!(m.source_tool, SourceTool::A1111);
        assert_eq!(m.steps, Some(28));
        assert_eq!(m.cfg_scale, Some(7.0));
        assert_eq!(m.seed, Some(12345));
        assert_eq!(m.sampler_name.as_deref(), Some("DPM++ 2M"));
        assert_eq!(m.model_name.as_deref(), Some("animagineXL"));
        assert_eq!(m.negative_prompt.as_deref(), Some("bad hands"));
        assert_eq!(m.loras.len(), 1);
        assert_eq!(m.loras[0].name, "detail");
        assert_eq!(m.loras[0].weight, 0.8);
        assert_eq!(m.model_type.unwrap().as_str(), "sdxl");
    }
}
